### ISSUE TITLE
Qwen3.5-35B-A3B VLM 训练速度很慢，如何利用 grouped_mm / Megatron EP 加速？

### ISSUE THREAD

**Issue by ChaosCodes** [modelscope/ms-swift#8145]
## 环境

- **GPU**: 8× A800 80GB (单机)
- **模型**: Qwen3.5-35B-A3B（VLM，包含 vision encoder）
- **ms-swift**: 4.0.0.dev0（从 main 分支安装）
- **megatron-core**: 0.15.0
- **PyTorch**: 2.7
- **训练类型**: full（全参数微调），VLM SFT
- **数据**: 自定义 VLM 数据集（~826K 样本，单图 + 文本）

## 相关 Issue / PR

- #8101 — Qwen3.5-397B-A17B 训练速度很慢（同样的问题，官方回复说 linear attention 目前用 transformers 实现，正在切 megatron-core）
- #8126 — [megatron] qwen3.5 use megatron-core（正在开发中的 PR，将 `Qwen3_5MoeGatedDeltaNet` 切换到 megatron-core 实现）
- #8142 — Qwen3.5-35B-A3B LoRA SFT loss=0 bug
- #8129 — Qwen3.5 megatron 训练报错 IndexError

## 问题描述

我们在 8×A800 上对 Qwen3.5-35B-A3B 做 VLM SFT，速度非常慢。我们尝试了两个框架：

### 1. verl FSDP2（baseline）

| 设置 | 值 |
|------|------|
| 并行方式 | FSDP2 (fully_shard) |
| micro_batch_size | 4 |
| global_batch_size | 256 |
| gradient_accumulation | 8 |
| max_length | 2048 |
| attn_implementation | eager |
| **s/step** | **~750s** |
| **tokens/s** | **~699** |

### 2. ms-swift + DeepSpeed ZeRO-3 + grouped_mm

```bash
NPROC_PER_NODE=8 swift sft \
    --model Qwen3.5-35B-A3B \
    --tuner_type full \
    --experts_impl grouped_mm \
    --per_device_train_batch_size 1 \
    --gradient_accumulation_steps 32 \
    --deepspeed zero3 \
    --max_length 2048
```

| 设置 | 值 |
|------|------|
| micro_batch_size | 1（mbs=4 + ZeRO-2 OOM） |
| global_batch_size | 256 |
| gradient_accumulation | 32 |
| **s/step** | **~1504s**（step 1） |
| **tokens/s** | **~348** |
| 显存/卡 | ~78GB |

ms-swift + ZeRO-3 比 verl FSDP2 还慢了约 2 倍，有 memory pressure 导致 cache flush。

### 对比参考

同样的 8×A800 环境，Qwen3-VL-8B（dense）可以达到 **~57,268 tokens/s**，Qwen3.5 只有其 1/82。

## 想请教的问题

1. **PR #8126 进度**：看到 @Jintao-Huang 在 #8101 提到会将 `Qwen3_5MoeGatedDeltaNet` 切成 megatron-core 实现。请问这个 PR 预计什么时候 merge？merge 后是否直接支持 **VLM**（带 vision encoder）的训练？

2. **grouped_mm 效果**：我们开了 `--experts_impl grouped_mm`，但在 ZeRO-3 全参数微调下速度反而更慢。ZeRO-3 下 grouped_mm 是否能正常加速？

3. **Megatron EP 路径**：`examples/models/qwen3_5/mcore.sh` 使用了 `expert_model_parallel_size=4` + `moe_grouped_gemm` + `moe_shared_expert_overlap`。这条路径在 8×A800 单机上做 **VLM** 全参数微调是否可行？需要 megatron-core 0.16 吗？

4. **当前可用的优化**：在等 #8126 merge 之前，有没有什么立即可用的加速方法？比如 `SWIFT_PATCH_CONV3D=1`、或者特定的 deepspeed 配置？

5. **flash attention**：Qwen3.5 的 linear attention（GatedDeltaNet）是否有 flash attention 或类似的优化实现？目前在 verl 中被迫使用 `eager`，这可能是速度慢的一个重要原因。

## 期望

希望能在 8×A800 上将 Qwen3.5-35B-A3B 的训练吞吐提升到至少 **2000-5000 tokens/s** 的水平（目前只有 ~700 tokens/s）。

感谢！

**Comment by Jintao-Huang** [comment]
稍等 在排查了

**Comment by ga546** [comment]
> 稍等 在排查了
I have the same problem. Is there a solution? Thank you very much.

**Comment by Jintao-Huang** [comment]
Try uninstalling fla. In my testing, it's faster without installing fla.

**Comment by KeuntekLee** [comment]
> Try uninstalling fla. In my testing, it's faster without installing fla.

What about flash_attn? Does it impact on Qwen3.5 series?

**Comment by ga546** [comment]
> Try uninstalling fla. In my testing, it's faster without installing fla.

Thanks for your reply. I tried it, and it does get faster, but the speed difference compared to the Qwen3-VL-32B is still significant. With the same amount of data and the same GPU, it's about four times slower.

**Comment by Jintao-Huang** [comment]
I'm still testing and will provide a benchmark.

**Comment by Jintao-Huang** [comment]
https://github.com/fla-org/flash-linear-attention/blob/d792da6dd544d28980205817b20535500c3f243e/fla/modules/l2norm.py#L162

It seems that this function is being invoked, leading to kernel recompilation whenever the sequence length varies.

**Comment by Jintao-Huang** [comment]
https://github.com/fla-org/flash-linear-attention/issues/758

**Comment by Jintao-Huang** [comment]
After fixing it locally, the speed is back to normal.

<img width="930" height="285" alt="Image" src="https://github.com/user-attachments/assets/c3100a20-3ec0-4c0a-99f3-6f330fd26812" />

```shell
PYTORCH_CUDA_ALLOC_CONF='expandable_segments:True' \
NPROC_PER_NODE=4 \
CUDA_VISIBLE_DEVICES=4,5,6,7 \
megatron sft \
    --model Qwen/Qwen3.5-35B-A3B \
    --save_safetensors true \
    --merge_lora true \
    --dataset 'AI-ModelScope/LongAlpaca-12k' \
    --load_from_cache_file true \
    --add_non_thinking_prefix true \
    --split_dataset_ratio 0.01 \
    --tuner_type lora \
    --lora_rank 8 \
    --lora_alpha 32 \
    --target_modules all-linear \
    --expert_model_parallel_size 4 \
    --moe_permute_fusion true \
    --moe_grouped_gemm true \
    --moe_shared_expert_overlap true \
    --moe_aux_loss_coeff 1e-6 \
    --micro_batch_size 1 \
    --global_batch_size 16 \
    --recompute_granularity full \
    --recompute_method uniform \
    --recompute_num_layers 1 \
    --num_train_epochs 1 \
    --group_by_length true \
    --finetune true \
    --freeze_llm false \
    --freeze_vit true \
    --freeze_aligner true \
    --cross_entropy_loss_fusion true \
    --lr 1e-4 \
    --lr_warmup_fraction 0.05 \
    --min_lr 1e-5 \
    --output_dir megatron_output/Qwen3.5-35B-A3B \
    --eval_steps 200 \
    --save_steps 200 \
    --max_length 8192 \
    --dataloader_num_workers 8 \
    --dataset_num_proc 8 \
    --no_save_optim true \
    --no_save_rng true \
    --sequence_parallel true \
    --attention_backend flash \
    --padding_free false
```

**Comment by KeuntekLee** [comment]
@Jintao-Huang Can u share ur local fixes for workaround?

**Comment by Jintao-Huang** [comment]
My fix was to remove the `NB` parameter from this file. Alternatively, you can use l2norm_fwd_kernel1 and l2norm_bwd_kernel1 as a workaround.

https://github.com/fla-org/flash-linear-attention/blob/d792da6dd544d28980205817b20535500c3f243e/fla/modules/l2norm.py#L85C5-L85C7

**Comment by Jintao-Huang** [comment]
According to the issue discussion, this can also be resolved by installing the FLA main branch.

https://github.com/fla-org/flash-linear-attention/issues/758

**Comment by Jintao-Huang** [comment]
这里新增了全参数训练的shell和速度截图

https://github.com/modelscope/ms-swift/pull/8194

**Comment by gaolongxi** [comment]
> 这里新增了全参数训练的shell和速度截图
> 
> [#8194](https://github.com/modelscope/ms-swift/pull/8194)

我在8*H100上面使用VQA数据训练Qwen3.5-2B时也遇到类似问题，相同数据集和卡数，3.5会比3慢5倍，20hours -> 100hours，安装的是fla的main branch，检查过issue中的两行代码，已经是修改后的版本。
我的超参数设置是
```
ARGS="
    --model ${MODEL_PATH} \
    --dataset ${DATASETS[@]} \
    --load_from_cache_file true \
    --add_non_thinking_prefix true \
    --loss_scale ignore_empty_think \
    --split_dataset_ratio 0.01 \
    --tuner_type full \
    --torch_dtype bfloat16 \
    --num_train_epochs ${EPOCHS} \
    --per_device_train_batch_size ${BATCH} \
    --per_device_eval_batch_size ${BATCH} \
    --learning_rate ${LR} \
    --gradient_accumulation_steps 1 \
    --gradient_checkpointing True \
    --group_by_length true \
    --freeze_llm false \
    --freeze_vit false \
    --freeze_aligner false \
    --output_dir ${OUTPUT_DIR} \
    --eval_steps 10000 \
    --save_steps 100000 \
    --save_total_limit 1 \
    --logging_steps 1 \
    --max_length 2048 \
    --warmup_ratio 0.05 \
    --dataloader_num_workers 8 \
    --dataset_num_proc 8 \
    --deepspeed /mnt/llmshared-ssd-hd/gaolongxi/projects/Qwen3-VL/qwen-vl-finetune/scripts/zero3_offload.json \
    --model_author swift \
    --model_name ${MODEL_NAME_LOWER}
"
```

**Comment by Jintao-Huang** [comment]
https://swift.readthedocs.io/en/latest/BestPractices/Qwen3_5-Best-Practice.html

causal_conv1d安装了嘛

**Comment by Jintao-Huang** [comment]
https://github.com/modelscope/ms-swift/blob/main/examples/models/qwen3_5/mcore_full.sh

**Comment by gaolongxi** [comment]
> https://swift.readthedocs.io/en/latest/BestPractices/Qwen3_5-Best-Practice.html
> 
> causal_conv1d安装了嘛

感谢回复！这个causal_conv1d是安装了的。您是推荐full finetune直接用megatron是吗
```

  Name                    Version       Build                 Channel    
───────────────────────────────────────────────────────────────────────────
  _openmp_mutex           4.5           20_gnu                conda-forge
  bzip2                   1.0.8         hda65f42_9            conda-forge
  ca-certificates         2026.2.25     hbd8a1cb_0            conda-forge
  causal_conv1d           1.6.0         pypi_0                pypi       
  deepspeed               0.18.6        pypi_0                pypi       
  flash-linear-attention  0.4.2         pypi_0                pypi       
  icu                     78.2          h33c6efd_0            conda-forge
  ld_impl_linux-64        2.45.1        default_hbd61a6d_101  conda-forge
  libexpat                2.7.4         hecca717_0            conda-forge
  libffi                  3.5.2         h3435931_0            conda-forge
  libgcc                  15.2.0        he0feb66_18           conda-forge
  libgcc-ng               15.2.0        h69a702a_18           conda-forge
  libgomp                 15.2.0        he0feb66_18           conda-forge
  liblzma                 5.8.2         hb03c661_0            conda-forge
  libnsl                  2.0.1         hb9d3cd8_1            conda-forge
  libsqlite               3.51.2        hf4e2dac_0            conda-forge
  libstdcxx               15.2.0        h934c35e_18           conda-forge
  libuuid                 2.41.3        h5347b49_0            conda-forge
  libxcrypt               4.4.36        hd590300_1            conda-forge
  libzlib                 1.3.1         hb9d3cd8_2            conda-forge
  liger_kernel            0.7.0         pypi_0                pypi       
  ms_swift                4.0.0         pypi_0                pypi       
  ncurses                 6.5           h2d0b736_3            conda-forge
  openssl                 3.6.1         h35e630c_1            conda-forge
  packaging               26.0          pyhcf101f3_0          conda-forge
  pip                     26.0.1        pyh8b19718_0          conda-forge
  python                  3.12.12       hd63d673_2_cpython    conda-forge
  qwen-vl-utils           0.0.14        pypi_0                pypi       
  readline                8.3           h853b02a_0            conda-forge
  setuptools              82.0.0        pyh332efcf_0          conda-forge
  tk                      8.6.13        noxft_h366c992_103    conda-forge
  torch                   2.10.0+cu129  pypi_0                pypi       
  transformer_engine      2.12.0        pypi_0                pypi       
  transformers            5.2.0         pypi_0                pypi       
  tzdata                  2025c         hc9c84f9_1            conda-forge
  uv                      0.10.8        pypi_0                pypi       
  wheel                   0.46.3        pyhd8ed1ab_0          conda-forge
  zstd                    1.5.7         hb78ec9c_6            conda-forge
```
